### PR TITLE
[Bug fix / Feature] Allowing self.log within val callback functions. 

### LINKED ISSUE #3813: Calling module.log(...) within a callback fails
## 🐛 Bug

Calling pl_module.log(...) within a Callback fails, even though this is recommended by the documentation here: https://pytorch-lightning.readthedocs.io/en/latest/loggers.html#logging-from-a-callback

### Error

```
  File "my_callback_file.py", line XX, in on_validation_epoch_end
    pl_module.log_dict(my_metrics_dict)
  File "/home/local/USHERBROOKE/pain5474/opt/miniconda3/envs/cav/lib/python3.8/site-packages/pytorch_lightning/core/lightning.py", line 287, in log_dict
    self.log(
  File "/home/local/USHERBROOKE/pain5474/opt/miniconda3/envs/cav/lib/python3.8/site-packages/pytorch_lightning/core/lightning.py", line 233, in log
    self._results.log(
  File "/home/local/USHERBROOKE/pain5474/opt/miniconda3/envs/cav/lib/python3.8/site-packages/pytorch_lightning/core/step_result.py", line 171, in log
    self.__set_meta(
  File "/home/local/USHERBROOKE/pain5474/opt/miniconda3/envs/cav/lib/python3.8/site-packages/pytorch_lightning/core/step_result.py", line 217, in __set_meta
    _internal = self['meta']['_internal']
KeyError: '_internal'
python-BaseException
```
cc @nathanpainchaud 

This is happening on master

### Expected behavior

We can log from callbacks using the lightning module

### Environment

Happening on PyTorch Lightning master

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR should enable to log within callbacks.Callback functions containing val (`funcs_name = [f for f in dir(callbacks.Callback) if ("val" in f)]`)

Second PR will take care of adding the same bug fix / feature for `train` and `test`.

Partial step #4167 
Fixes #3813

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/trainer/logging/test_eval_loop_logging_1_0.py [hunk 1]
```diff
@@ -321,3 +321,62 @@ def test_monitor_val_epoch_end(tmpdir):
         checkpoint_callback=checkpoint_callback,
     )
     trainer.fit(model)
+
+def test_log_works_in_validation_callback(tmpdir):
+    """
+    Tests that log can be called within callback
+    """
+    import pytorch_lightning as pl
```
Comment (williamFalcon) [inline tests/trainer/logging/test_eval_loop_logging_1_0.py:329]: add this to import to the top of the file

File: tests/trainer/logging/test_eval_loop_logging_1_0.py [hunk 2]
```diff
@@ -321,3 +321,62 @@ def test_monitor_val_epoch_end(tmpdir):
...
+
+    os.environ['PL_DEV_DEBUG'] = '1'
+
+    funcs_name = [f for f in dir(callbacks.Callback) if ("val" in f)]
```
Comment (williamFalcon) [inline tests/trainer/logging/test_eval_loop_logging_1_0.py:333]: while this is very nice and concise, this will also be very hard to debug lol.
Can you just create a callback in this method that implements the hooks you want?

```python
class CB(Callback):
  def on_...():
     self.log(...)
  def on_...():
    self.log(...)
```

File: tests/trainer/logging/test_eval_loop_logging_1_0.py [hunk 3]
```diff
@@ -321,3 +321,62 @@ def test_monitor_val_epoch_end(tmpdir):
...
+
+            loss = self.validation_return_values[self.count]
+            self.count += 1
+            self.log("test_val_loss", loss)
+            return {"test_val_loss": loss}
+
+    max_epochs = 5
+    model = TestModel()
+    # using Early Stopping to make sure callback_metrics is correctly set for `on_validation_epoch_end` hook
+    early_stop_callback = pl.callbacks.EarlyStopping(monitor="test_val_loss", patience=3, verbose=True)
```
Comment (williamFalcon) [inline tests/trainer/logging/test_eval_loop_logging_1_0.py:365]: doesn't make sense to early stop on a test split haha. that's cheating ;)

Comment (tchaton) [reply tests/trainer/logging/test_eval_loop_logging_1_0.py]: You are right ! Changed it :)

File: tests/trainer/logging/test_eval_loop_logging_1_0.py (hunk 2 above)
Comment (tchaton) [reply tests/trainer/logging/test_eval_loop_logging_1_0.py]: Done ! Yes, you are right. I was trying something 👍

File: pytorch_lightning/trainer/connectors/logger_connector.py [hunk 4]
```diff
@@ -171,39 +173,40 @@ def _log_on_evaluation_epoch_end_metrics(self, epoch_logs):
...
 
-            # actually log
-            if len(logger_metrics) > 0:
-                metrics_to_log.append(logger_metrics)
+                # actually log
+                if len(logger_metrics) > 0:
+                    metrics_to_log.append(logger_metrics)
 
-        # log all the metrics as a s single dict
+    def log_on_evaluation_end(self, metrics_to_log):
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/connectors/logger_connector.py:206]: this needs to be more specific

log_on_evaluation_end (step or epoch?)

File: tests/trainer/logging/test_eval_loop_logging_1_0.py [hunk 5]
```diff
@@ -321,3 +322,88 @@ def test_monitor_val_epoch_end(tmpdir):
...
+        num_sanity_val_steps=0,
+        max_epochs=max_epochs,
+        callbacks=[early_stop_callback, test_callback]
+    )
+    trainer.fit(model)
+
+    expected_logged_metrics = set(test_callback.callback_funcs_called + ["c", "val_loss"])
+    logged_metrics = set(trainer.logged_metrics.keys())
+    assert logged_metrics == expected_logged_metrics, logged_metrics
+    assert trainer.current_epoch == expected_stop_epoch
```
Comment (williamFalcon) [inline tests/trainer/logging/test_eval_loop_logging_1_0.py:409]: can you also assert the number of times the metrics were called per hook name?

Comment (williamFalcon) [reply tests/trainer/logging/test_eval_loop_logging_1_0.py:409]: this is also conflating two tests: early stopping and logging.
Can you separate both tests to keep things simple?

One test for the new changes to the metrics.
One test for early stopping (is this even necessary? don't we have a ton of these tests already?)

File: tests/models/test_onnx.py [hunk 6]
```diff
@@ -127,7 +132,7 @@ def test_error_if_input_sample_is_not_tensor(tmpdir):
                                          f'`Tensor`'):
         model.to_onnx(file_path, input_sample)
 
-
+@pytest.mark.skipif(not HAS_ONNX_RUNTIME, reason="onnxruntime wrongly configured")
```
Comment (williamFalcon) [inline tests/models/test_onnx.py:135]: what does onnx have to do with logging metrics?
can this be a separate PR?

File: pytorch_lightning/trainer/connectors/logger_connector.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector.py:206]: Yes :) Changed it for `log_epoch_metrics_on_evaluation_end`.

Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector.py:206]: Please, switch to PR: https://github.com/PyTorchLightning/pytorch-lightning/pull/4215/files

File: tests/trainer/logging/test_eval_loop_logging_1_0.py (hunk 5 above)
Comment (tchaton) [reply tests/trainer/logging/test_eval_loop_logging_1_0.py:409]: Yes, I let me remove it. It was to make sure the callback_metrics was properly set. Mainly for debugging purpose.
Also, there is an inconsistency with on_batch_start, on_batch_end

File: tests/models/test_onnx.py (hunk 6 above)
Comment (tchaton) [reply tests/models/test_onnx.py:135]: onnx doesn't install properly on both Sean and I computers for some reasons. And I was breaking Discovery Tests process. So I added this skip if import fails. I will open a PR for it apart this one :)
